pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,371 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificate
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	certutil "k8s.io/client-go/util/cert"
+	clientcertificates "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/kubelet/util/csr"
+)
+
+const (
+	syncPeriod     = 1 * time.Hour
+	keyExtension   = ".key"
+	certExtension  = ".crt"
+	pairNamePrefix = "kubelet"
+	// shouldRotatePercent is the time remaining threshold for when the
+	// certificate should be rotated. It is calculated by using the difference
+	// of the certificate end date and the certificate start date and checking
+	// what percentage of that interval is left.
+	shouldRotatePercent = 10
+)
+
+// Manager maintains and updates the certificates in use by this kubelet. In
+// the background it communicates with the API server to get new certificates
+// for certificates about to expire.
+type Manager interface {
+	// Start the API server status sync loop.
+	Start()
+	// GetCertificate gets the current certificate from the certificate
+	// manager. This function matches the signature required by
+	// tls.Config.GetCertificate so it can be passed as TLS configuration. A
+	// TLS server will automatically call back here to get the correct
+	// certificate when establishing each new connection.
+	GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+type manager struct {
```

Comment:
Have a separate store component now. I didn't do something different for access yet because I'm going to look at solving for client certs next.